File: src/common/ai_client.py
"""Thin wrapper around the OpenAI SDK: model choice, retry/backoff, cost tracking."""

import json
import os

from openai import OpenAI
from tenacity import retry, stop_after_attempt, wait_random_exponential

DEFAULT_MODEL = os.environ.get("OPENAI_MODEL", "gpt-4o-mini")

# Rough per-1M-token prices (USD) for cost estimate logging. Update if pricing changes.
_PRICES = {"gpt-4o-mini": (0.15, 0.60)}


class AIClient:
    def __init__(self, model=DEFAULT_MODEL):
        self.model = model
        self.client = OpenAI()
        self.prompt_tokens = 0
        self.completion_tokens = 0
        self.calls = 0

    @retry(wait=wait_random_exponential(min=1, max=30), stop=stop_after_attempt(4))
    def _call(self, system_prompt, user_prompt):
        return self.client.chat.completions.create(
            model=self.model,
            response_format={"type": "json_object"},
            messages=[
                {"role": "system", "content": system_prompt},
                {"role": "user", "content": user_prompt},
            ],
        )

    def generate_json(self, system_prompt, user_prompt, required_keys):
        """Call the model and parse a JSON response. Returns dict or None if
        the response doesn't parse or is missing required keys."""
        response = self._call(system_prompt, user_prompt)
        self.calls += 1
        usage = response.usage
        if usage:
            self.prompt_tokens += usage.prompt_tokens
            self.completion_tokens += usage.completion_tokens

        content = response.choices[0].message.content
        try:
            data = json.loads(content)
        except (json.JSONDecodeError, TypeError):
            return None
        if not all(k in data for k in required_keys):
            return None
        return data

    def print_usage(self):
        cost = None
        if self.model in _PRICES:
            in_price, out_price = _PRICES[self.model]
            cost = (self.prompt_tokens / 1_000_000) * in_price + (
                self.completion_tokens / 1_000_000
            ) * out_price
        cost_str = f", est. cost ${cost:.4f}" if cost is not None else ""
        print(
            f"AI usage: {self.calls} calls, {self.prompt_tokens} prompt tokens, "
            f"{self.completion_tokens} completion tokens{cost_str}"
        )

File: lib/aiClient.ts
// Thin wrapper around OpenRouter's chat completions API (OpenAI-compatible):
// model choice, retry/backoff, cost tracking. Plain fetch — no SDK needed.

const DEFAULT_MODEL = process.env.OPENROUTER_MODEL || "openai/gpt-4o-mini";
const API_URL = "https://openrouter.ai/api/v1/chat/completions";

// Rough per-1M-token prices (USD), for cost estimate logging. Update if pricing changes.
const PRICES: Record<string, [number, number]> = { "openai/gpt-4o-mini": [0.15, 0.6] };

type OpenRouterResponse = {
  choices?: { message?: { content?: string } }[];
  usage?: { prompt_tokens?: number; completion_tokens?: number };
};

export class AIClient {
  private model: string;
  private apiKey: string;
  promptTokens = 0;
  completionTokens = 0;
  calls = 0;

  constructor(model = DEFAULT_MODEL) {
    this.model = model;
    const apiKey = process.env.OPENROUTER_API_KEY;
    if (!apiKey) throw new Error("The OPENROUTER_API_KEY environment variable is missing or empty.");
    this.apiKey = apiKey;
  }

  private async callWithRetry(systemPrompt: string, userPrompt: string, attempts = 4): Promise<OpenRouterResponse> {
    let lastErr: unknown;
    for (let i = 0; i < attempts; i++) {
      try {
        const res = await fetch(API_URL, {
          method: "POST",
          headers: {
            Authorization: `Bearer ${this.apiKey}`,
            "HTTP-Referer": "http://localhost:3000",
            "X-Title": "Local Business Lead Generator",
            "Content-Type": "application/json",
          },
          body: JSON.stringify({
            model: this.model,
            response_format: { type: "json_object" },
            messages: [
              { role: "system", content: systemPrompt },
              { role: "user", content: userPrompt },
            ],
          }),
        });
        if (!res.ok) throw new Error(`OpenRouter API error ${res.status}: ${await res.text()}`);
        return await res.json();
      } catch (err) {
        lastErr = err;
        await new Promise((r) => setTimeout(r, Math.min(30000, 1000 * 2 ** i) * (0.5 + Math.random())));
      }
    }
    throw lastErr;
  }

  async generateJson<T extends Record<string, unknown>>(
    systemPrompt: string,
    userPrompt: string,
    requiredKeys: string[]
  ): Promise<T | null> {
    const response = await this.callWithRetry(systemPrompt, userPrompt);
    this.calls++;
    if (response.usage) {
      this.promptTokens += response.usage.prompt_tokens ?? 0;
      this.completionTokens += response.usage.completion_tokens ?? 0;
    }

    const content = response.choices?.[0]?.message?.content;
    if (!content) return null;
    let data: Record<string, unknown>;
    try {
      data = JSON.parse(content);
    } catch {
      return null;
    }
    if (!requiredKeys.every((k) => k in data)) return null;
    return data as T;
  }

  usage() {
    const prices = PRICES[this.model];
    const cost = prices ? (this.promptTokens / 1_000_000) * prices[0] + (this.completionTokens / 1_000_000) * prices[1] : null;
    return { calls: this.calls, promptTokens: this.promptTokens, completionTokens: this.completionTokens, cost };
  }
}
